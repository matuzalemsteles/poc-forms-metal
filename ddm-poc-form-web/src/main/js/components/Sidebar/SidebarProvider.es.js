import Component from 'metal-jsx';
import {Config} from 'metal-state';
import { LayoutSupport } from 'ddm-poc-form-js-components/Layout/index.js';

class SidebarProvider extends Component {
    static PROPS = {
        fieldContext: Config.array(),

        context: Config.array(),

        fieldFocus: Config.object(),
    }

    static STATE = {
        fieldContext: Config.array(),
    }

    constructor(props, context) {
        super(props, context);

        this.state.fieldContext = props.fieldContext;
    }

    willReceiveProps(nextProps) {
        if (
            typeof nextProps.context !== 'undefined' &&
            nextProps.context.newVal.length
        ) {
            // this._syncContextWithFieldContext();
        }

        if (
            typeof nextProps.fieldContext !== 'undefined' && 
            nextProps.fieldContext.newVal.length
        ) {
            this.setState({
                fieldContext: nextProps.fieldContext.newVal
            });
        }
    }

    _syncContextWithFieldContext() {
        const { context, fieldFocus } = this.props;
        const { fieldContext } = this.state;
        const { indexColumn, indexPage, indexRow } = fieldFocus;

        const fieldSelected = LayoutSupport.getColumn(context, indexPage, indexRow, indexColumn)[0];

        const depth = (array) => {
            for (let i = 0; i < array.length; i++) {
                let item = array[i];

                if (item.key) {  
                    let propName = fieldSelected[item.key] || '';

                    item[item.key] = propName;
                }

                if (item.rows && item.rows.length) {
                    depth(item.rows);
                } else if (item.columns && item.columns.length) {
                    depth(item.columns);
                } else if (item.fields && item.fields.length) {
                    depth(item.fields);
                }
            }

            return array;
        }


        console.log(depth(fieldContext));
    }

    _handleShowChanged(event) {
        this.emit('showChanged', event);
    }

    render() {
        const { children } = this.props;
        const { fieldContext } = this.state;
        const Child = children[0];

        const events = {
            showChanged: this._handleShowChanged.bind(this),
            ...Child.props.events
        };

        Object.assign(Child.props, {...this.props}, events, fieldContext);

        return children;
    }
}

export default SidebarProvider;