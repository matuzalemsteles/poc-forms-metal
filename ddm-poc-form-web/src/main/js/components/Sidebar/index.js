import Sidebar from './Sidebar.es';
import SidebarProvider from './SidebarProvider.es';

export default Sidebar;
export {
    Sidebar,
    SidebarProvider
}