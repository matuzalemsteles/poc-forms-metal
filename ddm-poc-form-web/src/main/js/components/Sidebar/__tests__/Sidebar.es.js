import {dom as MetalTestUtil} from 'metal-dom';
import {JSXComponent} from 'metal-jsx';
import createElement from './__mock__/createElement.es';
import Sidebar from '../Sidebar.es';

let component;
let spritemap = 'icons.svg';

const fieldLists = [
    {
        name: 'Date',
        type: 'date',
        icon: 'calendar',
        description: 'Select date from a Datepicker.'
    },
    {
        name: 'Text Field',
        type: 'text',
        icon: 'text',
        description: 'Single line or multiline text area.'
    },
    {
        name: 'Single Selection',
        type: 'radio',
        icon: 'radio-button',
        description: 'Select only one item with a radio button.'
    },
    {
        name: 'Select from list',
        type: 'select',
        icon: 'list',
        description: 'Choose an or more options from a list.'
    },
    {
        name: 'Grid',
        type: 'grid',
        icon: 'grid',
        description: 'Select options from a matrix.'
    },
    {
        name: 'Multiple Selection',
        type: 'checkbox',
        icon: 'select-from-list',
        description: 'Select multiple options using a checkbox.'
    }
];

const fieldContext = [
    {
        rows: [
            {
                columns: [
                    {
                        fields: [
                            {
                                key: 'label',
                                type: 'text',
                                label: 'Label',
                                spritemap,
                            }
                        ],
                        size: 12
                    }
                ]
            }
        ]
    }
];

describe('Sidebar', () => {
    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(() => {
        if (component) {
            component.dispose();
        }
    });

    it('should render the default markup', () => {
        component = new Sidebar({
            spritemap
        });

        expect(component).toMatchSnapshot();
    });

    it('should render a Sidebar open', () => {
        component = new Sidebar({
            spritemap
        });
        component.show();

        jest.runAllTimers();

        expect(component).toMatchSnapshot();
    });

    it('should render a Sidebar closed', () => {
        component = new Sidebar({
            spritemap
        });
        component.show();
        component.close();

        jest.runAllTimers();

        expect(component).toMatchSnapshot();
    });

    it('should render a Sidebar with fieldLists', () => {
        component = new Sidebar({
            spritemap,
            fieldLists,
        });

        expect(component).toMatchSnapshot();
    });

    it('should render a Sidebar with spritemap', () => {
        component = new Sidebar({
            spritemap,
        });

        expect(component).toMatchSnapshot();
    });

    it('should open Sidebar when is edit mode', () => {
        component = new Sidebar({
            fieldFocus: {
                indexColumn: 0,
                indexPage: 0,
                indexRow: 0,
                type: 'text'
            },
            fieldLists,
            fieldContext,
            spritemap
        });

        const spy = jest.spyOn(component, 'show');

        component.props.mode = 'edit';

        jest.runAllTimers();

        expect(component.state.show).toBeTruthy();
        expect(spy).toHaveBeenCalled();
    });

    it('should not update the internal mode with the above mode changes if are not in edit mode', () => {
        component = new Sidebar({
            spritemap,
        });

        const spy = jest.spyOn(component, '_setMode');

        component.props.mode = 'edit';

        jest.runAllTimers();

        expect(component.state.mode).toBe('add');
        expect(spy).toHaveBeenCalled();
    });

    it('should update the internal mode with the above mode changes when in edit mode', () => {
        component = new Sidebar({
            fieldFocus: {
                indexColumn: 0,
                indexPage: 0,
                indexRow: 0,
                type: 'text'
            },
            fieldLists,
            fieldContext,
            spritemap
        });

        const spy = jest.spyOn(component, '_setMode');

        component.props.mode = 'edit';

        jest.runAllTimers();

        expect(component.state.mode).toBe(component.props.mode);
        expect(spy).toHaveBeenCalled();
    });

    it('should reset drag and drop when context changes', () => {
        component = new Sidebar({
            spritemap,
        });

        const spyDrag = jest.spyOn(component._dragAndDrop, 'disposeInternal');
        const spy = jest.spyOn(component, '_startDrag');

        component.props.context = [
            {
                title: 'Untitled page'
            }
        ];

        jest.runAllTimers();

        expect(spy).toHaveBeenCalled();
        expect(spyDrag).toHaveBeenCalled();
    });

    it('should continue to propagate the fieldEdit event', () => {
        component = new Sidebar({
            mode: 'edit',
            fieldFocus: {
                indexColumn: 0,
                indexPage: 0,
                indexRow: 0,
                type: 'text'
            },
            fieldLists,
            fieldContext,
            spritemap
        });
        
        const spy = jest.spyOn(component, 'emit');
        const { layoutRenderer } = component.refs;

        layoutRenderer.emit('fieldEdit', {});

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
            'fieldEdit',
            expect.any(Object)
        );
    });

    it('should close sidebar when dragging an item', () => {
        component = new Sidebar({
            fieldLists,
            spritemap
        });

        component.show();

        expect(component.state.show).toBeTruthy();

        component._handleDrag(jest.fn());

        jest.runAllTimers();

        expect(component.state.show).toBeFalsy();
    });

    it('should emit a fieldAdd event when adding in layout', () => {
        component = new Sidebar({
            fieldLists,
            spritemap
        });

        const spy = jest.spyOn(component, 'emit');
        const { field1 } = component.refs;
        const element = createElement({
            tagname: 'div',
            attributes: [
                {
                    key: 'data-ddm-field-column',
                    value: 0
                },
                {
                    key: 'data-ddm-field-row',
                    value: 2
                },
                {
                    key: 'data-ddm-field-page',
                    value: 2
                }
            ]
        });
        const mockEvent = {
            target: {
                parentElement: element,
            },
            source: field1
        };

        component._handleFieldMove(mockEvent, {
            preventDefault: jest.fn()
        });

        jest.runAllTimers();

        expect(spy).toHaveBeenCalled();
        expect(spy).toHaveBeenCalledWith(
            'fieldAdd',
            {
                target: {
                    indexColumn: 0,
                    indexPage: 2,
                    indexRow: 2
                },
                fieldProperties: expect.any(Object),
                data: expect.anything()
            }
        )
    });

    it('should not emit when there is no target to drag item', () => {
        component = new Sidebar({
            fieldLists,
            spritemap
        });

        const spy = jest.spyOn(component, 'emit');
        const mockEvent = {
            target: undefined,
            source: undefined,
        };

        component._handleFieldMove(mockEvent, {
            preventDefault: jest.fn()
        });

        expect(spy).not.toHaveBeenCalled();
    });

    describe('Edit mode', () => {
        it('should not go into edit mode with just fieldContext', () => {
            component = new Sidebar({
                spritemap,
                fieldContext,
            });
    
            expect(component).toMatchSnapshot();
        });
    
        it('should not enter editing mode with only `edit` in mode', () => {
            component = new Sidebar({
                spritemap,
                mode: 'edit'
            });

            expect(component._isEditMode()).toBeFalsy();
            expect(component).toMatchSnapshot();
        });

        it('should go into edit mode with just fieldContext, fieldLists and fieldFocus', () => {
            component = new Sidebar({
                mode: 'edit',
                fieldFocus: {
                    indexColumn: 0,
                    indexPage: 0,
                    indexRow: 0,
                    type: 'text'
                },
                fieldLists,
                fieldContext,
                spritemap
            });
    
            jest.runAllTimers();
    
            expect(component).toMatchSnapshot();
        });
    
        it('should return true when there is fieldFocus, edit mode, fieldLists, and fieldContext', () => {
            component = new Sidebar({
                mode: 'edit',
                fieldFocus: {
                    indexColumn: 0,
                    indexPage: 0,
                    indexRow: 0,
                    type: 'text'
                },
                fieldLists,
                fieldContext,
                spritemap,
            });
    
            expect(component._isEditMode()).toBe(true);
        });

        it('should return false when there is only fieldFocus', () => {
            component = new Sidebar({
                spritemap,
                fieldFocus: {
                    indexColumn: 0,
                    indexPage: 0,
                    indexRow: 0,
                    type: 'text'
                }
            });
    
            expect(component._isEditMode()).toBe(false);
        });
    });

    describe('Interaction with markup', () => {
        it('should the close Sidebar when click outside Sidebar', () => {
            component = new Sidebar({
                spritemap,
            });

            const spy = jest.spyOn(component, '_handleDocClick');

            component.show();

            MetalTestUtil.triggerEvent(document, 'click', {});

            jest.runAllTimers();

            expect(component.state.show).toBeFalsy();
            expect(spy).toHaveBeenCalled();
        });

        it('should not close Sidebar when click inside Sidebar', () => {
            component = new Sidebar({
                spritemap,
            });

            const spy = jest.spyOn(component, '_handleDocClick');

            component.show();

            MetalTestUtil.triggerEvent(component.element, 'click', {});

            jest.runAllTimers();

            expect(component.state.show).toBeTruthy();
            expect(spy).toHaveBeenCalled();
        });

        it('should close Sidebar when click the button close', () => {
            component = new Sidebar({
                spritemap,
            });

            component.show();

            expect(component.state.show).toBeTruthy();

            const spy = jest.spyOn(component, 'close');
            const { close } = component.refs;

            close.click();

            jest.runAllTimers();

            expect(component.state.show).toBeFalsy();
            expect(spy).toHaveBeenCalled();
        });

        it('should change the tab on edit mode', () => {
            component = new Sidebar({
                mode: 'edit',
                fieldFocus: {
                    indexColumn: 0,
                    indexPage: 0,
                    indexRow: 0,
                    type: 'text'
                },
                fieldLists,
                fieldContext,
                spritemap,
            });

            const { tab1 } = component.refs;

            MetalTestUtil.triggerEvent(tab1, 'click', {});

            jest.runAllTimers();

            expect(component.state.tabActive).toBe(1);
        });

        it('should return to add mode', () => {
            component = new Sidebar({
                mode: 'edit',
                fieldFocus: {
                    indexColumn: 0,
                    indexPage: 0,
                    indexRow: 0,
                    type: 'text'
                },
                fieldLists,
                fieldContext,
                spritemap,
            });

            const { previousButton } = component.refs;

            MetalTestUtil.triggerEvent(previousButton.element, 'click', {});

            jest.runAllTimers();

            expect(component.state.mode).toBe('add');
        });
    });
});