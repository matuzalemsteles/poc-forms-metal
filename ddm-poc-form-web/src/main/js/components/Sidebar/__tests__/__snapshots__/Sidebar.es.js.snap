// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Edit mode should go into edit mode with just fieldContext, fieldLists and fieldFocus 1`] = `
<div class="sidebar-container" ref="sidebar">
  <div class="sidebar sidebar-light">
    <nav class="component-tbar tbar">
      <div class="container-fluid">
        <ul class="tbar-nav">
          <li class="tbar-item">
            <button class="btn btn-sm btn-secondary" aria-label="angle-left" type="button">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-angle-left" focusable="false">
                <title>angle-left</title>
                <use xlink:href="icons.svg#angle-left"></use>
              </svg>
            </button>
          </li>
          <li class="tbar-item tbar-item-expand text-left">
            <div>
              <button class="btn btn-sm btn-secondary" disabled="disabled" type="button">
                <span class="inline-item inline-item-before">
                  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-text" focusable="false">
                    <title>text</title>
                    <use xlink:href="icons.svg#text"></use>
                  </svg>
                </span>Text Field</button>
            </div>
          </li>
          <li class="tbar-item">
            <a class="component-action" href="#1" role="button" ref="close">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                <use xlink:href="icons.svg#times"></use>
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
      <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
        <span class="navbar-text-truncate">Details</span>
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="icons.svg#caret-bottom"></use>
        </svg>
      </a>
      <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
        <ul class="nav navbar-nav" role="tablist">
          <li ref="tab0" class="nav-item">
            <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
              <span class="navbar-text-truncate">Basic</span>
            </a>
          </li>
          <li ref="tab1" class="nav-item">
            <a class="nav-link" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
              <span class="navbar-text-truncate">Properties</span>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div class="ddm-sidebar-body">
      <div class="sidebar-body">
        <div class="tab-content">
          <div aria-labelledby="sidebarLightFormsFieldTab" class="fade tab-pane active show" id="sidebarLightDetails" role="tabpanel">
            <div class="row">
              <div class="col-md-12"></div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Sidebar Edit mode should not enter editing mode with only \`edit\` in mode 1`] = `
<div class="sidebar-container" ref="sidebar">
  <div class="sidebar sidebar-light">
    <nav class="component-tbar tbar">
      <div class="container-fluid">
        <ul class="tbar-nav">
          <li class="tbar-item tbar-item-expand text-left">
            <div class="tbar-section">
              <span class="text-truncate-inline">
                <span class="text-truncate">Add Elements</span>
              </span>
            </div>
          </li>
          <li class="tbar-item">
            <a class="component-action" href="#1" role="button" ref="close">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                <use xlink:href="icons.svg#times"></use>
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
      <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
        <span class="navbar-text-truncate">Details</span>
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="icons.svg#caret-bottom"></use>
        </svg>
      </a>
      <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
        <ul class="nav navbar-nav" role="tablist">
          <li ref="tab0" class="nav-item">
            <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
              <span class="navbar-text-truncate">Elements</span>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div class="ddm-sidebar-body"></div>
  </div>
</div>
`;

exports[`Sidebar Edit mode should not go into edit mode with just fieldContext 1`] = `
<div class="sidebar-container" ref="sidebar">
  <div class="sidebar sidebar-light">
    <nav class="component-tbar tbar">
      <div class="container-fluid">
        <ul class="tbar-nav">
          <li class="tbar-item tbar-item-expand text-left">
            <div class="tbar-section">
              <span class="text-truncate-inline">
                <span class="text-truncate">Add Elements</span>
              </span>
            </div>
          </li>
          <li class="tbar-item">
            <a class="component-action" href="#1" role="button" ref="close">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                <use xlink:href="icons.svg#times"></use>
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
      <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
        <span class="navbar-text-truncate">Details</span>
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="icons.svg#caret-bottom"></use>
        </svg>
      </a>
      <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
        <ul class="nav navbar-nav" role="tablist">
          <li ref="tab0" class="nav-item">
            <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
              <span class="navbar-text-truncate">Elements</span>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div class="ddm-sidebar-body"></div>
  </div>
</div>
`;

exports[`Sidebar should render a Sidebar closed 1`] = `
<div class="sidebar-container" ref="sidebar">
  <div class="sidebar sidebar-light">
    <nav class="component-tbar tbar">
      <div class="container-fluid">
        <ul class="tbar-nav">
          <li class="tbar-item tbar-item-expand text-left">
            <div class="tbar-section">
              <span class="text-truncate-inline">
                <span class="text-truncate">Add Elements</span>
              </span>
            </div>
          </li>
          <li class="tbar-item">
            <a class="component-action" href="#1" role="button" ref="close">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                <use xlink:href="icons.svg#times"></use>
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
      <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
        <span class="navbar-text-truncate">Details</span>
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="icons.svg#caret-bottom"></use>
        </svg>
      </a>
      <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
        <ul class="nav navbar-nav" role="tablist">
          <li ref="tab0" class="nav-item">
            <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
              <span class="navbar-text-truncate">Elements</span>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div class="ddm-sidebar-body"></div>
  </div>
</div>
`;

exports[`Sidebar should render a Sidebar open 1`] = `
<div class="sidebar-container show" ref="sidebar">
  <div class="sidebar sidebar-light">
    <nav class="component-tbar tbar">
      <div class="container-fluid">
        <ul class="tbar-nav">
          <li class="tbar-item tbar-item-expand text-left">
            <div class="tbar-section">
              <span class="text-truncate-inline">
                <span class="text-truncate">Add Elements</span>
              </span>
            </div>
          </li>
          <li class="tbar-item">
            <a class="component-action" href="#1" role="button" ref="close">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                <use xlink:href="icons.svg#times"></use>
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
      <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
        <span class="navbar-text-truncate">Details</span>
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="icons.svg#caret-bottom"></use>
        </svg>
      </a>
      <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
        <ul class="nav navbar-nav" role="tablist">
          <li ref="tab0" class="nav-item">
            <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
              <span class="navbar-text-truncate">Elements</span>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div class="ddm-sidebar-body"></div>
  </div>
</div>
`;

exports[`Sidebar should render a Sidebar with fieldLists 1`] = `
<div class="sidebar-container" ref="sidebar">
  <div class="sidebar sidebar-light">
    <nav class="component-tbar tbar">
      <div class="container-fluid">
        <ul class="tbar-nav">
          <li class="tbar-item tbar-item-expand text-left">
            <div class="tbar-section">
              <span class="text-truncate-inline">
                <span class="text-truncate">Add Elements</span>
              </span>
            </div>
          </li>
          <li class="tbar-item">
            <a class="component-action" href="#1" role="button" ref="close">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                <use xlink:href="icons.svg#times"></use>
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
      <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
        <span class="navbar-text-truncate">Details</span>
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="icons.svg#caret-bottom"></use>
        </svg>
      </a>
      <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
        <ul class="nav navbar-nav" role="tablist">
          <li ref="tab0" class="nav-item">
            <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
              <span class="navbar-text-truncate">Elements</span>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div class="ddm-sidebar-body">
      <ul class="list-group">
        <li class="list-group-header">
          <h3 class="list-group-header-title">Basic Elements</h3>
        </li>
        <div ref="field0" data-ddm-field-type-index="0" class="ddm-drag-item list-group-item list-group-item-flex">
          <div class="autofit-col">
            <div class="sticker sticker-secondary">
              <svg class="lexicon-icon lexicon-icon-calendar" aria-hidden="true">
                <use xlink:href="icons.svg#calendar"></use>
              </svg>
            </div>
          </div>
          <div class="autofit-col autofit-col-expand">
            <h4 class="list-group-title text-truncate">
              <span>Date</span>
            </h4>
            <p class="list-group-subtitle text-truncate">Select date from a Datepicker.</p>
          </div>
        </div>
        <div ref="field1" data-ddm-field-type-index="1" class="ddm-drag-item list-group-item list-group-item-flex">
          <div class="autofit-col">
            <div class="sticker sticker-secondary">
              <svg class="lexicon-icon lexicon-icon-text" aria-hidden="true">
                <use xlink:href="icons.svg#text"></use>
              </svg>
            </div>
          </div>
          <div class="autofit-col autofit-col-expand">
            <h4 class="list-group-title text-truncate">
              <span>Text Field</span>
            </h4>
            <p class="list-group-subtitle text-truncate">Single line or multiline text area.</p>
          </div>
        </div>
        <div ref="field2" data-ddm-field-type-index="2" class="ddm-drag-item list-group-item list-group-item-flex">
          <div class="autofit-col">
            <div class="sticker sticker-secondary">
              <svg class="lexicon-icon lexicon-icon-radio-button" aria-hidden="true">
                <use xlink:href="icons.svg#radio-button"></use>
              </svg>
            </div>
          </div>
          <div class="autofit-col autofit-col-expand">
            <h4 class="list-group-title text-truncate">
              <span>Single Selection</span>
            </h4>
            <p class="list-group-subtitle text-truncate">Select only one item with a radio button.</p>
          </div>
        </div>
        <div ref="field3" data-ddm-field-type-index="3" class="ddm-drag-item list-group-item list-group-item-flex">
          <div class="autofit-col">
            <div class="sticker sticker-secondary">
              <svg class="lexicon-icon lexicon-icon-list" aria-hidden="true">
                <use xlink:href="icons.svg#list"></use>
              </svg>
            </div>
          </div>
          <div class="autofit-col autofit-col-expand">
            <h4 class="list-group-title text-truncate">
              <span>Select from list</span>
            </h4>
            <p class="list-group-subtitle text-truncate">Choose an or more options from a list.</p>
          </div>
        </div>
        <div ref="field4" data-ddm-field-type-index="4" class="ddm-drag-item list-group-item list-group-item-flex">
          <div class="autofit-col">
            <div class="sticker sticker-secondary">
              <svg class="lexicon-icon lexicon-icon-grid" aria-hidden="true">
                <use xlink:href="icons.svg#grid"></use>
              </svg>
            </div>
          </div>
          <div class="autofit-col autofit-col-expand">
            <h4 class="list-group-title text-truncate">
              <span>Grid</span>
            </h4>
            <p class="list-group-subtitle text-truncate">Select options from a matrix.</p>
          </div>
        </div>
        <div ref="field5" data-ddm-field-type-index="5" class="ddm-drag-item list-group-item list-group-item-flex">
          <div class="autofit-col">
            <div class="sticker sticker-secondary">
              <svg class="lexicon-icon lexicon-icon-select-from-list" aria-hidden="true">
                <use xlink:href="icons.svg#select-from-list"></use>
              </svg>
            </div>
          </div>
          <div class="autofit-col autofit-col-expand">
            <h4 class="list-group-title text-truncate">
              <span>Multiple Selection</span>
            </h4>
            <p class="list-group-subtitle text-truncate">Select multiple options using a checkbox.</p>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Sidebar should render a Sidebar with spritemap 1`] = `
<div class="sidebar-container" ref="sidebar">
  <div class="sidebar sidebar-light">
    <nav class="component-tbar tbar">
      <div class="container-fluid">
        <ul class="tbar-nav">
          <li class="tbar-item tbar-item-expand text-left">
            <div class="tbar-section">
              <span class="text-truncate-inline">
                <span class="text-truncate">Add Elements</span>
              </span>
            </div>
          </li>
          <li class="tbar-item">
            <a class="component-action" href="#1" role="button" ref="close">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                <use xlink:href="icons.svg#times"></use>
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
      <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
        <span class="navbar-text-truncate">Details</span>
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="icons.svg#caret-bottom"></use>
        </svg>
      </a>
      <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
        <ul class="nav navbar-nav" role="tablist">
          <li ref="tab0" class="nav-item">
            <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
              <span class="navbar-text-truncate">Elements</span>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div class="ddm-sidebar-body"></div>
  </div>
</div>
`;

exports[`Sidebar should render the default markup 1`] = `
<div class="sidebar-container" ref="sidebar">
  <div class="sidebar sidebar-light">
    <nav class="component-tbar tbar">
      <div class="container-fluid">
        <ul class="tbar-nav">
          <li class="tbar-item tbar-item-expand text-left">
            <div class="tbar-section">
              <span class="text-truncate-inline">
                <span class="text-truncate">Add Elements</span>
              </span>
            </div>
          </li>
          <li class="tbar-item">
            <a class="component-action" href="#1" role="button" ref="close">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
                <use xlink:href="icons.svg#times"></use>
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <nav class="component-navigation-bar navbar navigation-bar navbar-collapse-absolute navbar-expand-md navbar-underline">
      <a aria-controls="sidebarLightCollapse00" aria-expanded="false" aria-label="Toggle Navigation" class="collapsed navbar-toggler navbar-toggler-link" data-toggle="collapse" href="#sidebarLightCollapse00" role="button">
        <span class="navbar-text-truncate">Details</span>
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="icons.svg#caret-bottom"></use>
        </svg>
      </a>
      <div class="collapse navbar-collapse" id="sidebarLightCollapse00">
        <ul class="nav navbar-nav" role="tablist">
          <li ref="tab0" class="nav-item">
            <a class="nav-link active" aria-controls="sidebarLightDetails" data-toggle="tab" href="#" role="tab">
              <span class="navbar-text-truncate">Elements</span>
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div class="ddm-sidebar-body"></div>
  </div>
</div>
`;
